+ newrelic: add possibility to send akka metrics to the newrelic

slavaschmidt opened the pull request:
Added possibility to send akka metrics to the newrelic as custom metrics.
Externalized categories of newrelic subscription into the configuration file.
This allow to define which metrics categories should be send to newrelic
independent upon which metrics are actually collected.

Akka metrics exported to the newrelic as custom metrics and available in custom
dashboards in some format similar to:

Cusom/akka-actor/{ActorSystemName[user|system|...]\{ActorName}/{MetricName}

for actors and

Custom/akka-thread-pool-executor\Some-Service\akka.io.pinned-dispatcher/ProcessedTasks

for thread pools.

Same metrics for multiple actors can be displayed as a single chart by using *
(star) for part of the actor name. For example

Cusom/akka-actor\MyActor\user\*DatabaseWorker/ProcessingTime

will show processing time for all database workers.

Example of actor metrics displayed by newrelic:
http://s4.postimg.org/sfn9vjzgt/Screen_Shot_2015_04_22_at_11_24_15.png

Example of pool metrics displayed by newrelic:
http://s4.postimg.org/gchy7zoel/Screen_Shot_2015_04_22_at_11_24_24.png

dpsoft commented:
@slavaschmidt  thanks for this PR!!! and LGTM. What you think @ivantopo?.

ivantopo commented on kamon-newrelic/src/main/resources/reference.conf:
```diff
@@ -23,6 +23,18 @@ kamon {
+
+    subscriptions {
+      trace           = [ "**" ]
+      trace-segment   = [ "**" ]
```
I think that trace and trace-segment subscriptions should not be configurable or at least not in the same place. Those two categories are the very foundation of how the New Relic dashboard works and should not be treated the same as we treat custom metrics.

We could always subscribe in code to all traces and segments and change this key to `custom-metric-subscriptions` to hold all the customized additions. That would also feel more natural to the users as they will know for sure that whatever they put in those keys will be reported as custom metrics. 

If you want to go the extra mile, the trace and trace-segment subscriptions could be also configurable via a separate config key, but at the moment I don't find that to be particularly useful.

ivantopo commented on kamon-newrelic/src/main/scala/kamon/newrelic/CustomMetricExtractor.scala:
```diff
@@ -16,22 +16,31 @@
   def extract(settings: AgentSettings, collectionContext: CollectionContext, metrics: Map[Entity, EntitySnapshot]): Map[MetricID, MetricData] = {
-    def onlySimpleMetrics(kv: (Entity, EntitySnapshot)): Boolean =
-      kamon.metric.SingleInstrumentEntityRecorder.AllCategories.contains(kv._1.category)
+    val (simple, complex) = metrics.partition(simpleMetrics)
```
Here the `metrics` map will contain all metrics, included trace and trace-segment metrics, they should be filtered out as they will already be processed by the web transaction metrics extractor.

ivantopo commented:
Left a couple notes there! Looking forward to your feedback.

slavaschmidt commented:
Hi @ivantopo 
Many thanks for quick feedback. Sorry, i missed these SegmentMetrics and TraceMetrics subscriptions.
I've removed them from the config and added an additional filter in the subscription code.

ivantopo commented:
There is nothing to be sorry about! Now, I see you removed trace and trace-segment from the config file, but I don't see the code that is actually subscribing to those categories, am I missing something? Also, filtering them from the custom metrics extractor is also missing.. maybe you forgot to push some of those changes?

ivantopo commented on kamon-newrelic/src/main/scala/kamon/newrelic/Agent.scala:
```diff
@@ -143,4 +158,46 @@ object AgentSettings {
+
+  private val traceAndSegmentMetrics = Seq(TraceMetrics.category, SegmentMetrics.category)
+
+  def traceOrSegmentMetric(name: String) = traceAndSegmentMetrics contains name
```
`isTraceOrSegmentEntity` might be more appropriate, and pass in the `Entity` instead of just the category name. And please, add explicit return type! We know it is not necessary, but in general it makes code more readable :D.

ivantopo commented:
Closing this, as #228 was already merged and contains this changes.
